File: 31-next-permutation/31-next-permutation.cpp
class Solution {
public:
    void nextPermutation(vector<int>& nums) {
        return solve(nums);
    }
private:
    // Channa Sir
    // O(n logn)
    void solve(vector<int>& nums) {
        int n = nums.size();
        int i = n - 2;
        
        while(i >= 0 && nums[i] >= nums[i + 1])
            --i;
        
        if(i == -1) {
            reverse(nums.begin(), nums.end());
            return;
        }
        int j = i + 1;
        
        while(j < n && nums[j] > nums[i])
            ++j;
        --j;
        swap(nums[i], nums[j]);
        reverse(nums.begin() + i + 1, nums.end());
    }
};

File: 31-next-permutation/31-next-permutation.py
class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        n = len(nums)
        
        i = n - 2
        while i >= 0 and nums[i] >= nums[i + 1]:
            i -= 1
            
        if i != -1:
            j = i + 1
            while j < n and nums[j] > nums[i]:
                j += 1
                
            j -= 1
            nums[i], nums[j] = nums[j], nums[i]
        
        nums[i + 1:] = (nums[i + 1:])[::-1]